R/dependencies.R

```diff
@@ -0,0 +1,410 @@
+##' Then we find the lastest version of A
+##' get the artefact hashes from A
+##' match them against the depends hashes from B
+is_out_of_date <- function(con, child_id) {
+  ## filename - the name of the artefact B used (not the use name)
+  ## file_hash - the hash of the artefact B used
+  ## report_version - the id of the report where the artefact came from
+  ## is_pinned - was this report pinned to a speicif version?
+  ## is_latest - if the report was pinned was it to the lastest version
+  ## !NOTE! It is unclear what we do when both is_pinned and is_latest are true
+  chd_qry <- sprintf("depends.report_version='%s'", child_id)
+  sql_chd_qry <-
+    paste(c("SELECT", "filename,",
+                      "file_hash,",
+                      "is_pinned,",
+                      "is_latest,",
+                      "report_version_artefact.report_version",
+            "FROM", "(depends",
+            "INNER JOIN", "report_version", "ON",
+                          "depends.report_version=report_version.id",
+            "INNER JOIN", "file_artefact", "ON",
+                          "depends.use = file_artefact.id",
+            "INNER JOIN", "report_version_artefact", "ON",
+                          "report_version_artefact.id = file_artefact.artefact)",
+            "WHERE", chd_qry), collapse = " ")
+
+  db_chd <- DBI::dbGetQuery(con, sql_chd_qry)
+
+  ## this reports uses no artefacts - so can never be out of date
+  if (nrow(db_chd) == 0) {
+    return(FALSE)
+  }
+
+  ## we iterate over rows of the data frame and make sure the artefact match
+  for (i in 1:nrow(db_chd)) {
+    filename <- db_chd$filename[i]
+    file_hash <- db_chd$file_hash[i]
+    report_id <- db_chd$report_version[i]
+    ## we need to find the id latest version of the report with id = report_id
+    if (!db_chd$is_pinned[i]) { ## if not pinned...
+      latest_id <- get_latest_by_id(con, report_id) ## ..use latest
+    } else { ## if pinned to..
+      if (!db_chd$is_latest[i]) { ## ..not latest..
+        latest_id <- report_id ## ..use the pinned id
+      } else { ## pinned to latest
+        latest_id <- report_id
+      }
+    }
+
+
+    ## we return a dataframe with columns:
+    ## filename - the filename of the artefact
+    ## file_hash - the hash of the artefact
+    par_qry <- sprintf("report_version.id='%s'", latest_id)
+    sql_par_qry <-
+      paste(c("SELECT", "file_artefact.filename,",
+                        "file_artefact.file_hash",
+             "FROM", "(report_version",
+             "INNER JOIN", "report_version_artefact", "ON",
+                           "report_version.id = report_version_artefact.report_version",
+             "INNER JOIN", "file_artefact", "ON",
+                           "file_artefact.artefact = report_version_artefact.id)",
+             "WHERE", par_qry), collapse = " ")
+
+    db_par <- DBI::dbGetQuery(con, sql_par_qry)
+    i <- which(db_par$filename == filename)
+    if (db_par$file_hash[i] != file_hash)
```

Comment:
please always use braces, even for one-line statements. The `which` in the line above appears unneeded